File: 1. 단계별로_풀어보기/36. 트리에서의 동적 계획법/2. 1949.js
// 입력
const fs = require('fs')
const inputArr = fs
  .readFileSync(0)
  .toString()
  .trim()
  .split('\n')
  .map(row => row.trim())

// 입력 파싱
const N = +inputArr[0]
const peopleArr = inputArr[1].split(' ').map(Number)
const treeInfoArr = inputArr.slice(2).map(row => row.split(' ').map(Number))

const connInfoArr = Array.from({length: N + 1}, () => [])
const parentArr = Array.from({length: N + 1}).fill(-1)

const ROOT_NODE = 1

// 트리 연결정보 생성
treeInfoArr.forEach(info => {
  const [U, V] = info
  connInfoArr[U].push(V)
  connInfoArr[V].push(U)
})

parentArr[ROOT_NODE] = 0
const [k0, k1, k2] = DFS(ROOT_NODE)

process.stdout.write(`${Math.max(k0, k1, k2)}`)

function DFS(now) {
  /**
   * k0Arr: 자식노드가 선정되었을때의 최대인원의 배열
   * k1Arr: 자식노드가 1번연속 선정되지 못한 마을일때 최대인원의 배열
   * k2Arr: 자식노드가 2번연속 선정되지 못한 마을일때 최대인원의 배열
   */
  const k0Arr = []
  const k1Arr = []
  const k2Arr = []

  connInfoArr[now].forEach(next => {
    if (parentArr[next] === -1) {
      parentArr[next] = now
      const [k0, k1, k2] = DFS(next)

      k0Arr.push(k0)
      k1Arr.push(k1)
      k2Arr.push(k2)
    }
  })

  /**
   * 영역 1: nowK1 구하는곳.
   *
   * - 현재 미선정, 부모노드 미선정이어도 됨.
   * - 자식노드는 선정되었거나 1번연속 미선정이어야 함.
   *     + 자식이 2연속 미선정이면 현재노드는 무조건 선정되어야 함
   * - 자식노드중 최소 1개는 선정되어야함
   *
   */
  let nowK1 = 0
  let hasK0 = false
  for (let i = 0; i < k0Arr.length; i++) {
    nowK1 += Math.max(k0Arr[i], k1Arr[i])
    if (k0Arr[i] >= k1Arr[i]) hasK0 = true
  }
  if (!hasK0 && k0Arr.length > 0) {
    let maxDelta = -1000000000
    for (let i = 0; i < k0Arr.length; i++) {
      maxDelta = Math.max(maxDelta, k0Arr[i] - k1Arr[i])
    }
    nowK1 += maxDelta
  }

  /**
   * 영역 2: nowK2 구하는곳.
   *
   * - 현재 미선정, 부모노드 선정이어야 함.
   * - 자식노드는 선정되었거나 1번연속 미선정이어야 함.
   *     + 자식이 2연속 미선정이면 현재 노드는 무조건 선정되어야 함
   * - 자식노드중 최소 1개는 미선정이어야 함
   *     + 그래야 현재 노드가 2연속 미선정이 됨.
   */
  let nowK2 = 0
  for (let i = 0; i < k1Arr.length; i++) {
    nowK2 += k1Arr[i]
  }

  // 영역 3: nowK0 구하는곳. 현재 선정.
  // 자식노드는 전부 미선정이어야 함
  let nowK0 = peopleArr[now - 1]
  for (let i = 0; i < k1Arr.length; i++) {
    const k1 = k1Arr[i]
    const k2 = k2Arr[i]
    nowK0 += Math.max(k1, k2)
  }

  return [nowK0, nowK1, nowK2]
}

File: 1. 단계별로_풀어보기/36. 트리에서의 동적 계획법/2. 1949.py
import sys

sys.setrecursionlimit(100000)

input = sys.stdin.readline
output = sys.stdout.write

#######################
#      입력 영역       #
#######################

N = int(input())
peopleArr = list(map(int, input().split()))
connInfoArr = list([] for i in range(N + 1))

for _ in range(N - 1):
  U, V = map(int, input().split())
  connInfoArr[U].append(V)
  connInfoArr[V].append(U)

###############################
#      기타 자료구조 영역       #
###############################

ROOT_NODE = 1
parentArr = list(-1 for i in range(N + 1))
parentArr[ROOT_NODE] = ROOT_NODE

#############################
#      사용할 함수 영역       #
#############################

def DFS(now):
  k0_arr = []
  k1_arr = []
  k2_arr = []

  for next in connInfoArr[now]:
    if parentArr[next] == -1:
      parentArr[next] = now
      k0, k1, k2 = DFS(next)
      k0_arr.append(k0)
      k1_arr.append(k1)
      k2_arr.append(k2)

  lenArr = len(k0_arr)

  # 영역 0: 현재 마을은 선정되는 경우. 자식은 전부 미선정.
  nowK0 = peopleArr[now - 1]
  for i in range(lenArr):
    nowK0 += max(k1_arr[i], k2_arr[i])

  # 영역 1: 현재 마을은 1연속 미선정, 부모 마을도 미선정 가능한 경우.
  # 자식 중 최소 1개는 선정되어야 함. 자식은 k0 또는 k1만 가능(k2 불가).
  nowK1 = 0
  hasK0 = False
  for i in range(lenArr):
    nowK1 += max(k0_arr[i], k1_arr[i])
    if k0_arr[i] >= k1_arr[i]:
      hasK0 = True
  if not hasK0 and lenArr > 0:
    maxDelta = max(k0_arr[i] - k1_arr[i] for i in range(lenArr))
    nowK1 += maxDelta

  # 영역 2: 현재 마을은 2연속 미선정, 부모 마을은 선정되어야 함.
  # 자식 중 최소 1개는 미선정(k1)이어야 함. 자식은 k0 또는 k1만 가능.
  nowK2 = 0
  isVisitK1 = False
  for i in range(lenArr):
    if k0_arr[i] < k1_arr[i]:
      isVisitK1 = True
    nowK2 += max(k0_arr[i], k1_arr[i])
  if not isVisitK1 and lenArr > 0:
    minValue = min(k0_arr[i] - k1_arr[i] for i in range(lenArr))
    nowK2 -= minValue

  return [nowK0, nowK1, nowK2]


#######################
#      출력 영역       #
#######################

k0, k1, k2 = DFS(ROOT_NODE)
output(f"{max(k0, k1, k2)}")

